File: src/builtins/echo.cpp
// Implementation of the echo builtin.
#include "config.h"  // IWYU pragma: keep

#include "echo.h"

#include <climits>
#include <cstddef>

#include "../builtin.h"
#include "../common.h"
#include "../fallback.h"  // IWYU pragma: keep
#include "../io.h"
#include "../wgetopt.h"
#include "../wutil.h"  // IWYU pragma: keep

struct echo_cmd_opts_t {
    bool print_newline = true;
    bool print_spaces = true;
    bool interpret_special_chars = false;
};
static const wchar_t *const short_options = L"+:Eens";
static const struct woption *const long_options = nullptr;

static int parse_cmd_opts(echo_cmd_opts_t &opts, int *optind, int argc, const wchar_t **argv,
                          parser_t &parser, io_streams_t &streams) {
    UNUSED(parser);
    UNUSED(streams);
    const wchar_t *cmd = argv[0];
    int opt;
    wgetopter_t w;
    echo_cmd_opts_t oldopts = opts;
    int oldoptind = 0;
    while ((opt = w.wgetopt_long(argc, argv, short_options, long_options, nullptr)) != -1) {
        switch (opt) {
            case 'n': {
                opts.print_newline = false;
                break;
            }
            case 'e': {
                opts.interpret_special_chars = true;
                break;
            }
            case 's': {
                opts.print_spaces = false;
                break;
            }
            case 'E': {
                opts.interpret_special_chars = false;
                break;
            }
            case ':': {
                builtin_missing_argument(parser, streams, cmd, argv[w.woptind - 1]);
                return STATUS_INVALID_ARGS;
            }
            case '?': {
                opts = oldopts;
                *optind = w.woptind - 1;
                return STATUS_CMD_OK;
            }
            default: {
                DIE("unexpected retval from wgetopt_long");
            }
        }

        // Super cheesy: We keep an old copy of the option state around,
        // so we can revert it in case we get an argument like
        // "-n foo".
        // We need to keep it one out-of-date so we can ignore the *last* option.
        // (this might be an issue in wgetopt, but that's a whole other can of worms
        //  and really only occurs with our weird "put it back" option parsing)
        if (w.woptind == oldoptind + 2) {
            oldopts = opts;
            oldoptind = w.woptind;
        }
    }

    *optind = w.woptind;
    return STATUS_CMD_OK;
}

/// Parse a numeric escape sequence in str, returning whether we succeeded. Also return the number
/// of characters consumed and the resulting value. Supported escape sequences:
///
/// \0nnn: octal value, zero to three digits
/// \nnn: octal value, one to three digits
/// \xhh: hex value, one to two digits
static bool builtin_echo_parse_numeric_sequence(const wchar_t *str, size_t *consumed,
                                                unsigned char *out_val) {
    bool success = false;
    unsigned int start = 0;  // the first character of the numeric part of the sequence

    unsigned int base = 0, max_digits = 0;
    if (convert_digit(str[0], 8) != -1) {
        // Octal escape
        base = 8;

        // If the first digit is a 0, we allow four digits (including that zero); otherwise, we
        // allow 3.
        max_digits = (str[0] == L'0' ? 4 : 3);
    } else if (str[0] == L'x') {
        // Hex escape
        base = 16;
        max_digits = 2;

        // Skip the x
        start = 1;
    }

    if (base == 0) {
        return success;
    }

    unsigned int idx;
    unsigned char val = 0;  // resulting character
    for (idx = start; idx < start + max_digits; idx++) {
        int digit = convert_digit(str[idx], base);
        if (digit == -1) break;
        val = val * base + digit;
    }

    // We succeeded if we consumed at least one digit.
    if (idx > start) {
        *consumed = idx;
        *out_val = val;
        success = true;
    }
    return success;
}

/// The echo builtin.
///
/// Bash only respects -n if it's the first argument. We'll do the same. We also support a new,
/// fish specific, option -s to mean "no spaces".
maybe_t<int> builtin_echo(parser_t &parser, io_streams_t &streams, const wchar_t **argv) {
    const wchar_t *cmd = argv[0];
    UNUSED(cmd);
    int argc = builtin_count_args(argv);
    echo_cmd_opts_t opts;
    int optind;
    int retval = parse_cmd_opts(opts, &optind, argc, argv, parser, streams);
    if (retval != STATUS_CMD_OK) return retval;

    // The special character \c can be used to indicate no more output.
    bool continue_output = true;

    const wchar_t *const *args_to_echo = argv + optind;
    // We buffer output so we can write in one go,
    // this matters when writing to an fd.
    wcstring out;
    for (size_t idx = 0; continue_output && args_to_echo[idx] != nullptr; idx++) {
        if (opts.print_spaces && idx > 0) {
            out.push_back(' ');
        }

        const wchar_t *str = args_to_echo[idx];
        for (size_t j = 0; continue_output && str[j]; j++) {
            if (!opts.interpret_special_chars || str[j] != L'\\') {
                // Not an escape.
                out.push_back(str[j]);
            } else {
                // Most escapes consume one character in addition to the backslash; the numeric
                // sequences may consume more, while an unrecognized escape sequence consumes none.
                wchar_t wc;
                size_t consumed = 1;
                switch (str[j + 1]) {
                    case L'a': {
                        wc = L'\a';
                        break;
                    }
                    case L'b': {
                        wc = L'\b';
                        break;
                    }
                    case L'e': {
                        wc = L'\x1B';
                        break;
                    }
                    case L'f': {
                        wc = L'\f';
                        break;
                    }
                    case L'n': {
                        wc = L'\n';
                        break;
                    }
                    case L'r': {
                        wc = L'\r';
                        break;
                    }
                    case L't': {
                        wc = L'\t';
                        break;
                    }
                    case L'v': {
                        wc = L'\v';
                        break;
                    }
                    case L'\\': {
                        wc = L'\\';
                        break;
                    }
                    case L'c': {
                        wc = 0;
                        continue_output = false;
                        break;
                    }
                    default: {
                        // Octal and hex escape sequences.
                        unsigned char narrow_val = 0;
                        if (builtin_echo_parse_numeric_sequence(str + j + 1, &consumed,
                                                                &narrow_val)) {
                            // Here consumed must have been set to something. The narrow_val is a
                            // literal byte that we want to output (#1894).
                            wc = ENCODE_DIRECT_BASE + narrow_val % 256;
                        } else {
                            // Not a recognized escape. We consume only the backslash.
                            wc = L'\\';
                            consumed = 0;
                        }
                        break;
                    }
                }

                // Skip over characters that were part of this escape sequence (but not the
                // backslash, which will be handled by the loop increment.
                j += consumed;

                if (continue_output) {
                    out.push_back(wc);
                }
            }
        }
    }
    if (opts.print_newline && continue_output) {
        out.push_back('\n');
    }

    if (!out.empty()) {
        streams.out.append(out);
    }

    return STATUS_CMD_OK;
}

File: fish-rust/src/builtins/echo.rs
//! Implementation of the echo builtin.

use super::prelude::*;
use crate::wchar::encode_byte_to_char;

#[derive(Debug, Clone, Copy)]
struct Options {
    print_newline: bool,
    print_spaces: bool,
    interpret_special_chars: bool,
}

impl Default for Options {
    fn default() -> Self {
        Self {
            print_newline: true,
            print_spaces: true,
            interpret_special_chars: false,
        }
    }
}

fn parse_options(
    args: &mut [WString],
    parser: &Parser,
    streams: &mut IoStreams<'_>,
) -> Result<(Options, usize), Option<c_int>> {
    const SHORT_OPTS: &wstr = L!("+:Eens");
    const LONG_OPTS: &[woption] = &[];

    let mut opts = Options::default();

    let mut oldopts = opts;
    let mut oldoptind = 0;

    let mut w = wgetopter_t::new(SHORT_OPTS, LONG_OPTS, args);
    while let Some(c) = w.wgetopt_long() {
        match c {
            'n' => opts.print_newline = false,
            'e' => opts.interpret_special_chars = true,
            's' => opts.print_spaces = false,
            'E' => opts.interpret_special_chars = false,
            ':' => {
                builtin_missing_argument(parser, streams, w.cmd(), &w.argv()[w.woptind - 1], true);
                return Err(STATUS_INVALID_ARGS);
            }
            '?' => {
                return Ok((oldopts, w.woptind - 1));
            }
            _ => {
                panic!("unexpected retval from wgetopter::wgetopt_long()");
            }
        }

        // Super cheesy: We keep an old copy of the option state around,
        // so we can revert it in case we get an argument like
        // "-n foo".
        // We need to keep it one out-of-date so we can ignore the *last* option.
        // (this might be an issue in wgetopt, but that's a whole other can of worms
        //  and really only occurs with our weird "put it back" option parsing)
        if w.woptind == oldoptind + 2 {
            oldopts = opts;
            oldoptind = w.woptind;
        }
    }

    Ok((opts, w.woptind))
}

/// Parse a numeric escape sequence in `s`, returning the number of characters consumed and the
/// resulting value. Supported escape sequences:
///
/// - `0nnn`: octal value, zero to three digits
/// - `nnn`: octal value, one to three digits
/// - `xhh`: hex value, one to two digits
fn parse_numeric_sequence<I>(chars: I) -> Option<(usize, u8)>
where
    I: IntoIterator<Item = char>,
{
    let mut chars = chars.into_iter().peekable();

    // the first character of the numeric part of the sequence
    let mut start = 0;

    let mut base: u8 = 0;
    let mut max_digits = 0;

    let first = *chars.peek()?;
    if first.is_digit(8) {
        // Octal escape
        base = 8;

        // If the first digit is a 0, we allow four digits (including that zero); otherwise, we
        // allow 3.
        max_digits = if first == '0' { 4 } else { 3 };
    } else if first == 'x' {
        // Hex escape
        base = 16;
        max_digits = 2;

        // Skip the x
        start = 1;
    };

    if base == 0 {
        return None;
    }

    let mut val = 0;
    let mut consumed = start;
    for digit in chars
        .skip(start)
        .take(max_digits)
        .map_while(|c| c.to_digit(base.into()))
    {
        // base is either 8 or 16, so digit can never be >255
        let digit = u8::try_from(digit).unwrap();

        val = val * base + digit;

        consumed += 1;
    }

    // We succeeded if we consumed at least one digit.
    if consumed > 0 {
        Some((consumed, val))
    } else {
        None
    }
}

/// The echo builtin.
///
/// Bash only respects `-n` if it's the first argument. We'll do the same. We also support a new,
/// fish specific, option `-s` to mean "no spaces".
pub fn echo(parser: &Parser, streams: &mut IoStreams<'_>, args: &mut [WString]) -> Option<c_int> {
    let (opts, optind) = match parse_options(args, parser, streams) {
        Ok((opts, optind)) => (opts, optind),
        Err(err @ Some(_)) if err != STATUS_CMD_OK => return err,
        Err(err) => panic!("Illogical exit code from parse_options(): {err:?}"),
    };

    // The special character \c can be used to indicate no more output.
    let mut output_stopped = false;

    // We buffer output so we can write in one go,
    // this matters when writing to an fd.
    let mut out = WString::new();
    let args_to_echo = &args[optind..];
    'outer: for (idx, arg) in args_to_echo.iter().enumerate() {
        if opts.print_spaces && idx > 0 {
            out.push(' ');
        }

        let mut chars = arg.chars().peekable();
        while let Some(c) = chars.next() {
            if !opts.interpret_special_chars || c != '\\' {
                // Not an escape.
                out.push(c);
                continue;
            }

            let Some(next_char) = chars.peek() else {
                // Incomplete escape sequence is echoed verbatim
                out.push('\\');
                break;
            };

            // Most escapes consume one character in addition to the backslash; the numeric
            // sequences may consume more, while an unrecognized escape sequence consumes none.
            let mut consumed = 1;

            let escaped = match next_char {
                'a' => '\x07',
                'b' => '\x08',
                'e' => '\x1B',
                'f' => '\x0C',
                'n' => '\n',
                'r' => '\r',
                't' => '\t',
                'v' => '\x0B',
                '\\' => '\\',
                'c' => {
                    output_stopped = true;
                    break 'outer;
                }
                _ => {
                    // Octal and hex escape sequences.
                    if let Some((digits_consumed, narrow_val)) =
                        parse_numeric_sequence(chars.clone())
                    {
                        consumed = digits_consumed;
                        // The narrow_val is a literal byte that we want to output (#1894).
                        encode_byte_to_char(narrow_val)
                    } else {
                        consumed = 0;
                        '\\'
                    }
                }
            };

            // Skip over characters that were part of this escape sequence (after the backslash
            // that was consumed by the `while` loop).
            // TODO: `Iterator::advance_by()`: https://github.com/rust-lang/rust/issues/77404
            for _ in 0..consumed {
                let _ = chars.next();
            }

            out.push(escaped);
        }
    }

    if opts.print_newline && !output_stopped {
        out.push('\n');
    }

    if !out.is_empty() {
        streams.out.append(&out);
    }

    STATUS_CMD_OK
}
